### ISSUE TITLE
Gradient penalty with mixed precision training

### ISSUE THREAD

**Issue by RocaVincent** [tensorflow/tensorflow#48662]
**System information**
- TensorFlow version (you are using): 2.5.0
- Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state.**
I haven't found a way to implement a stable gradient penalty with mixed precision. I think it comes from the use of second order gradient (the penalty is computed inside the overall loss GradientTape context). Whether I scale the gradient penalty or not (avoiding underflow), the global loss scale quickly decreases to 1 during training. I don't know if this feature necessarily requires a new TensorFlow feature, but at least a tip not mentionned in [the mixed precision guide](https://www.tensorflow.org/guide/mixed_precision).

**Will this change the current api? How?**
No

**Who will benefit with this feature?**
At least every programmer who would like to implement a Wassertein GAN with gradient penalty, which is now the most validated setup for GAN training actually.

**Comment by bhack** [comment]
Is this the same as https://github.com/tensorflow/tensorflow/issues/47156 ?

**Comment by reedwm** [comment]
I am unfamiliar with gradient penalty. @nluehr @benbarsdell do you know how to use gradient penalty with loss scaling?

@bhack, as you mentioned, #47156 might be related. Given gradient penalty involves second order gradients, we should determine how to apply loss scaling when using gradient penalty.

**Comment by RocaVincent** [comment]
@bhack Like @reedwm saied, #47156 is related. Indeed, it seems that I don't have problems when I put all instance normalization layers in float32 precision (I should have said this in the first post). But the problem is also linked to the use of gradient penalty, because with other models, I manage to keep some instance normalization layers in mixed precision without having the loss scale decreasing to 1. Therefore, I think the issue is relevant and as @reedwm saied, it would be good to determine how to manage second order gradients in mixed precision training.

**Comment by bhack** [comment]
If I see the example [WGAN_WP in keras](https://keras.io/examples/generative/wgan_gp/) they use a custom `train_step`.
Are the functions in mixed precision [custom_training_loop](https://www.tensorflow.org/guide/mixed_precision?hl=en#training_the_model_with_a_custom_training_loop) section sufficent?

**Comment by RocaVincent** [comment]
Yes I saw both sites but the keras one doesn't use mixed precision and the tensorflow guide doesn't talk about scaling with second order gradient.

**Comment by bhack** [comment]
So are you looking to something like https://pytorch.org/docs/stable/notes/amp_examples.html#gradient-penalty?
They are unscaling the grad_param to add the penalty to the loss.

**Comment by RocaVincent** [comment]
That's what I do now. But I don't manage to transpose well the Pytorch example you give in Tensorflow because it seems there isn't a GradientTape equivalent in the code. In my case, I have a GradientTape context which is used to compute the gradients for the gradient penalty inside an other GradientTape which computes the global loss as it is in [the Keras example you linked](https://keras.io/examples/generative/wgan_gp/). So, I don't see how I could adapt this Pytorch example to my case.

**Comment by bhack** [comment]
So this is your code example in March right? https://stackoverflow.com/questions/66771093/tensorflow-mixed-precision-second-order-gradient-scaling

**Comment by RocaVincent** [comment]
Yes it is.

**Comment by reedwm** [comment]
In general, loss scaling can be applied as by scaling the loss and unscaling the gradients around every call to `tape.gradient` (the methods `Model.fit`, `Optimizer.minimize` and `Optimizer.get_gradients` do this for you automatically but for the examples ahead, I will use `tape.gradient`). For example, suppose gradient penalty is implemented as follows:

```python
with tf.GradientTape() as outer_tape:
  with tf.GradientTape() as inner_tape:
    output = model(x)
    loss = loss_fn(output)
  grads = inner_tape.gradient(loss, weights)
  grad_norm = tf.add_n([tf.reduce_sum(g ** 2) for g in grads])
  grad_norm = tf.sqrt(grad_norm)
  loss += grad_norm
grads = outer_tape.gradient(loss, weights)
opt.apply_gradients(list(zip(grads, weights)))
```

To apply loss scaling, before each call to `tape.gradient`, the loss must be scaled and after each call, the gradients must be unscaled. In the example below, each added line has a commented added to help indicate what has changed:

```python
opt = tf.keras.mixed_precision.LossScaleOptimizer(opt)         # Wrap opt in LSO
with tf.GradientTape() as outer_tape:
  with tf.GradientTape() as inner_tape:
    output = model(x)
    loss = loss_fn(output)
    scaled_loss = opt.get_scaled_loss(loss)                     # Scale loss                 
  scaled_gradients = inner_tape.gradient(scaled_loss, weights)      
  gradients = opt.get_unscaled_gradients(scaled_gradients)      # Unscale grads
  grad_norm = tf.add_n([tf.reduce_sum(g ** 2) for g in gradients])
  grad_norm = tf.sqrt(grad_norm)
  loss += grad_norm
  scaled_loss = opt.get_scaled_loss(loss)                       # Scale loss
scaled_gradients = outer_tape.gradient(scaled_loss, weights)
gradients = opt.get_unscaled_gradients(scaled_gradients)        # Unscale grads
opt.apply_gradients(list(zip(gradients, weights)))
```

This is the same as when computing single-order gradients, except that `tape.gradient` is called twice instead of once so the loss and gradients need to be scaled twice instead of once. I am still not familiar on gradient penalty and why it is done, but following the example above will ensure every tensor in an intermediate gradient is scaled exactly once (as what must always be done with any model when using mixed precision). This also matches what the PyTorch example does that was linked by @bhack above.

**Comment by RocaVincent** [comment]
@reedwm That's what I do currently and this implementation gives me problems with instance normalization layers in mixed-precision mode whereas I don't have these problems with same architectures but without gradient penalty. But maybe I have to admit I can't use instance normalization layers in mixed precision with gradient penalty, hoping I will not have problems with other mixed precision layers.

**Comment by reedwm** [comment]
In general, computing variance in float16 can cause issues since it may overflow or underflow, so it should always be computed in float32. Instance normalization computes variance, so either the variance within instance normalization must be done in float32 or the entire instance normalization itself must be float32.

I'm not sure why you only run into issues with mixed precision instance normalization when using gradient penalty. Perhaps gradient penalty causes the variance to be more likely to overflow/underflow in float16.

**Comment by bhack** [comment]
I've not tried It but probably you can take a look how mixed precision and gp was handled in https://github.com/bryanlimy/tf2-cyclegan

**Comment by RocaVincent** [comment]
@reedwn Yes I know the overflow issue in normalization layers due to mean and variance computing. But make all normalization calculations in float32 is very expansive in terms of memory and time. So I tried to keep a maximum of computation in float16. To be more precise, I observed which normalization layers in my models were likely to overflow by first running in full precision and checking values (checking if sum is in the float16 range) when computing a mean. Here is my layer code : 

```python
class InstanceNormalization(Layer):
    
    def __init__(self, epsilon=1e-3, mean_f32=False, mean_abs_dev_f32=False, **kwargs):
        super().__init__(**kwargs)
        self.epsilon=epsilon
        self.mean_f32 = mean_f32
        self.mean_abs_dev_f32 = mean_abs_dev_f32
        
    def build(self, batch_input_shape):
        self.scale = self.add_weight(
            name='scale',
            shape=batch_input_shape[-1:],
            initializer=random_normal_initializer(1,0.02),
            trainable=True
        )
        self.offset = self.add_weight(
            name='offset',
            shape=batch_input_shape[-1:],
            initializer="zeros",
            trainable=True
        )
        self.axis = range(1, len(batch_input_shape)-1)
        super().build(batch_input_shape)
        
    def call(self, x):
        if self.mean_f32:
            mean = tf_cast(k_mean(tf_cast(x,"float32"), axis=self.axis, keepdims=True), "float16")
        else:
            mean = k_mean(x, axis=self.axis, keepdims=True)
        dev = x-mean
        if self.mean_abs_dev_f32:
            mean_abs_dev = tf_cast(k_mean(tf_cast(k_abs(dev), "float32"), axis=self.axis, keepdims=True),"float16")
        else:
            mean_abs_dev = k_mean(k_abs(dev), axis=self.axis, keepdims=True)
        normalized = dev / (mean_abs_dev+self.epsilon)
        return self.scale*normalized + self.offset
```
As you see in the `call` method, the two if/else control if mean computations are done in float32 or float16.


@bhack Yes I already saw this code and it seems that they don't use scaling/unscaling for the inner gradient tape. In addition, they use tensorflow normalization layers and I think computations are all done in float32 even if we set the global policy to mixed precision.

**Comment by reedwm** [comment]
By using average absolute deviation instead of standard deviation, you remove the risk of overflow, so I think float16 is OK.

If you are concerned about performance of running variance calculations in float32, consider [using XLA](https://www.tensorflow.org/xla). XLA can potentially remove the performance cost of doing the variance in float32, since it can fuse the cast-to-float16, the variance calculations, and the cast-to-float32. Also, XLA is faster in general and so will help performance in most cases.

**Comment by RocaVincent** [comment]
@reedwm Yes I already use XLA and indeed it saves a little bit of memory and computation time.

**Comment by RocaVincent** [comment]
As I feared about, it seems that problem occur even without float16 instance normalization (IN) layers. I tried running WGAN-GP with only float32 IN layers and the training was not good. I quickly got inf values and like before, the loss scale parameter quickly goes down to 1. Thus, I think the problem of mixed precision with second order gradient is a quite general problem.

**Comment by tilakrayal** [comment]
Hi,

The computing variance in float16 can cause issues since it may overflow or underflow, so it should always be computed in float32.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
